### PR TITLE
R4R: Add HTLC module and implement HTLC creation

### PR DESCRIPTION
**Pull request by secret2830**
Resolves: https://github.com/irisnet/irishub/issues/1946

### CODE REVIEW

File: app/v2/htlc/internal/keeper/keeper.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+	createTags := sdk.NewTags(
+		types.TagSender, []byte(htlc.Sender),
+		types.TagReceiver, []byte(htlc.Sender),
+		types.TagReceiverOnOtherChain, []byte(htlc.ReceiverOnOtherChain),
+		types.TagOutAmount, []byte(htlc.OutAmount.String()),
+		types.TagInAmount, sdk.Uint64ToBigEndian(htlc.InAmount),
+		types.TagSecretHashLock, []byte(hex.EncodeToString(htlc.GetSecretHashLock())),
+		types.TagTimestamp, sdk.Uint64ToBigEndian(htlc.Timestamp),
+		types.TagExpireHeight, sdk.Uint64ToBigEndian(htlc.ExpireHeight),
```
Comment (chengwenxi) [inline app/v2/htlc/internal/keeper/keeper.go]: Too many tags, I think only `sender`,`receiver`, `secretHashLock` is necessary.

File: app/v2/htlc/internal/keeper/keeper_keys.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+// KeyHTLCTokens returns the key for the HTLC-locked token by the given token id
+func KeyHTLCTokens(tokenId string) []byte {
+	return append(PrefixHTLCToken, []byte(tokenId)...)
+}
+
+// KeyHTLCExpireQueue returns the key prefix for HTLC expiration queue
+func KeyHTLCExpireQueue(expireHeight uint64, secretHashLock []byte) []byte {
+	return append([]byte(fmt.Sprintf("PrefixHTLCExpireQueue:%d:", expireHeight)), secretHashLock...)
```
Comment (chengwenxi) [inline app/v2/htlc/internal/keeper/keeper_keys.go]: The prefix will be sorted as `PrefixHTLCExpireQueue:1`,`PrefixHTLCExpireQueue:10`,`PrefixHTLCExpireQueue:2`..., not 1,2...10.

File: app/v2/htlc/internal/keeper/keeper.go (hunk 1 above)
Comment (SegueII) [reply app/v2/htlc/internal/keeper/keeper.go]: I think, if want to generate HTLC on another chain automatically with tools, these tags are necessary.

File: app/v2/htlc/internal/types/htlc.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	"fmt"
+
+	sdk "github.com/irisnet/irishub/types"
+)
+
+// HTLC represents a HTLC
+type HTLC struct {
+	Sender               sdk.AccAddress `json:"sender"`                 // the initiator address
+	Receiver             sdk.AccAddress `json:"receiver"`               // the recipient address
+	ReceiverOnOtherChain []byte         `json:"receiver_in_other_htlc"` // the recipient address on other chain
```
Comment (haifengxi) [inline app/v2/htlc/internal/types/htlc.go]: `receiver_in_other_htlc` --> `receiver_on_other_chain`

**Pull request merged**